src/crypto/qat/qcccrypto.cc

```diff
@@ -322,150 +327,279 @@ bool QccCrypto::perform_op(unsigned char* out, const unsigned char* in,
+    do {
+      status = cpaCySymDpInitSession(cyInstHandle,
+                                  &sessionSetupData,
+                                  *sessionCtx);
+    } while (CPA_STATUS_RETRY == status);
+    if (status != CPA_STATUS_SUCCESS) {
+      dout(1) << "cpaCySymDpInitSession failed with status = " << status << dendl;
     }
+  } else {
+    dout(1) << "Session alloc failed with status = " << status << dendl;
   }
-  pthread_join(cypollthreads[avail_inst], NULL);
+  return status;
+}
 
-  if(qcc_op_mem[avail_inst].op_result != CPA_STATUS_SUCCESS) {
-    derr << "Unable to perform crypt operation" << dendl;
-    return false;
-  }
+bool QccCrypto::symPerformOp(int avail_inst,
+                              CpaCySymSessionCtx sessionCtx,
+                              const Cpa8U *pSrc,
+                              Cpa8U *pDst,
+                              Cpa32U chunk_size,
+                              Cpa32U total_len,
+                              Cpa8U *pIv,
+                              Cpa32U ivLen) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  Cpa32U one_batch_size = chunk_size * MAX_NUM_SYM_REQ_BATCH;
+  for (Cpa32U off = 0; off < total_len; off += one_batch_size) {
+    CompletionHandle complete;
+    for (Cpa32U offset = off, i = 0; offset < total_len && i < MAX_NUM_SYM_REQ_BATCH; offset += chunk_size, i++) {
+      CpaCySymDpOpData *pOpData = qcc_op_mem[avail_inst].sym_op_data[i];
+      Cpa8U *pSrcBuffer = qcc_op_mem[avail_inst].src_buff[i];
+      Cpa8U *pIvBuffer = qcc_op_mem[avail_inst].iv_buff[i];
+      Cpa32U process_size = offset + chunk_size <= total_len ? chunk_size : total_len - offset;
+      if (CPA_STATUS_SUCCESS == status) {
+        // copy source into buffer
+        memcpy(pSrcBuffer, pSrc + offset, process_size);
+        // copy IV into buffer
+        memcpy(pIvBuffer, &pIv[i * ivLen], ivLen);
+      }
+      complete.add_one();
+      if (CPA_STATUS_SUCCESS == status) {
+        //pOpData assignment
+        pOpData->thisPhys = qaeVirtToPhysNUMA(pOpData);
+        pOpData->instanceHandle = qcc_inst->cy_inst_handles[avail_inst];
+        pOpData->sessionCtx = sessionCtx;
+        pOpData->pCallbackTag = (void *)&complete;
+        pOpData->cryptoStartSrcOffsetInBytes = 0;
+        pOpData->messageLenToCipherInBytes = process_size;
+        pOpData->iv = qaeVirtToPhysNUMA(pIvBuffer);
+        pOpData->pIv = pIvBuffer;
+        pOpData->ivLenInBytes = ivLen;
+        pOpData->srcBuffer = qaeVirtToPhysNUMA(pSrcBuffer);
+        pOpData->srcBufferLen = process_size;
+        pOpData->dstBuffer = qaeVirtToPhysNUMA(pSrcBuffer);
+        pOpData->dstBufferLen = process_size;
+      }
+
+      if (CPA_STATUS_SUCCESS == status) {
+        do {
+          status = cpaCySymDpEnqueueOp(pOpData, CPA_FALSE);
+        } while (CPA_STATUS_RETRY == status);
+        if (CPA_STATUS_SUCCESS != status) {
+          dout(1) << "cpaCySymDpEnqueueOp failed. (status = " << status << ")" << dendl;
```

Comment:
which case cause enqueue failed? and how to handle enqueue failed ?

Reply:
if SAL is failed to run in crypto data plane, it will return an error. So if instance start normally, we may not meet the fail. And keep the fail condition to collect the error we don't know. And if the error happened, it will return false to rgw crypto and submit the requests that enqueued to clean the queue.
And that only no space is available on the queue will return CPA_STATUS_RETRY.  There we use 32 request in one batch and default config is 64 for asym and 512 for sym in QAT config, so this can be omitted.